src/persist/src/indexed/mod.rs

```diff
@@ -325,54 +560,91 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
                 }
             }
         }
+        Ok(())
+    }
 
-        let prev = self.serialize_meta();
-        let mut updates_by_id: HashMap<Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>> = HashMap::new();
-        for (id, updates) in updates.into_iter() {
-            updates_by_id.entry(id).or_default().extend(updates);
+    /// Asynchronously persists (Key, Value, Time, Diff) updates for the stream
+    /// with the given id.
+    pub fn write(
+        &mut self,
+        updates: Vec<(Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>)>,
+        res: FutureHandle<SeqNo>,
+    ) {
+        if let Err(err) = self.write_sync_validate(&updates) {
+            res.fill(Err(err));
+            return;
         }
 
-        // Give each write a unique, incrementing sequence number, and use
-        // futures_seqno_upper to track the sequence number of the next write.
         let write_seqno = self.futures_seqno_upper;
-        self.futures_seqno_upper = SeqNo(write_seqno.0 + 1);
+        self.pending_writes.extend(updates);
+        self.pending_responses
+            .push(PendingResponse::SeqNo(res, write_seqno));
+    }
+
+    /// Flush pending writes to future.
+    ///
+    /// The caller is responsible for committing metadata after
+    /// this, and then flushing pending responses.
+    fn flush_pending_writes_to_future(&mut self) -> Result<(), Error> {
+        let mut updates_by_id: HashMap<Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>> = HashMap::new();
+        for (id, updates) in self.pending_writes.drain(..) {
+            updates_by_id.entry(id).or_default().extend(updates);
+        }
 
-        // This range represents the [lower, upper) of sequence numbers assigned
-        // to this write.
-        //
-        // TODO: do we still need sequence numbers? This will make more sense
-        // when we send multiple writes to future at once but I'm not sure if
-        // we need the concept of sequence numbers when we're not reading from
-        // a buffer. On the other hand, how would we distinguish future batches
-        // from each other?
-        let desc = write_seqno..self.futures_seqno_upper;
         let updates_for_listeners = updates_by_id.clone();
-        for (id, updates) in updates_by_id.drain() {
-            let future = self
-                .futures
-                .get_mut(&id)
-                .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
+        if !updates_by_id.is_empty() {
```

Comment:
i had that at first but its not actually correct because even if pending writes are empty we still need to commit metadata if e.g. we have modified seal. ill restructuring the code to split the actual dealing with pending writes part into its own function or something else to make this clearer

Reply:
restructured into a separate function